lib/active_merchant/billing/gateways/cardprocess.rb

```diff
@@ -0,0 +1,252 @@
+module ActiveMerchant #:nodoc:
+  module Billing #:nodoc:
+    class CardprocessGateway < Gateway
+      self.test_url = 'https://test.vr-pay-ecommerce.de/v1/payments'
+      self.live_url = 'https://vr-pay-ecommerce.de/v1/payments'
+
+      self.supported_countries = ['US']
+      self.default_currency = 'EUR'
+      self.supported_cardtypes = [:visa, :master, :american_express, :diners_club, :jcb]
+
+      self.homepage_url = 'https://vr-pay-ecommerce.docs.oppwa.com/'
+      self.display_name = 'CardProcess VR-Pay'
+      self.money_format = :dollars
+
+      STANDARD_ERROR_CODE_MAPPING = {}
+
+      # Creates a new CardProcess Gateway
+      #
+      # The gateway requires a valid login, password, and entity ID
+      # to be passed in the +options+ hash.
+      #
+      # === Options
+      #
+      # * <tt>:user_id</tt> -- The CardProcess user ID
+      # * <tt>:password</tt> -- The CardProcess password
+      # * <tt>:entity_id</tt> -- The CardProcess channel or entity ID for any transactions
+      def initialize(options={})
+        requires!(options, :user_id, :password, :entity_id)
+        super
+        # This variable exists purely to allow remote tests to force error codes;
+        # the lack of a setter despite its usage is intentional.
+        @extra_options = {}
+      end
+
+      def purchase(money, payment, options = {})
+        post = {}
+        add_invoice(post, money, options)
+        add_payment(post, payment)
+        add_address(post, payment, options)
+        add_customer_data(post, options)
+
+        commit('DB', post)
```

Comment:
I might consider mapping these action codes to readable aliases.

Reply:
I actually originally had that, but it felt a bit weird to me, since each alias was used only once, in the appropriate function, at which point it didn't seem to be getting much. It looked like this, roughly:

```ruby
ACTIONS = {
  purchase: 'DB',
  ...
}

def purchase(...)
  ...
  commit(ACTIONS[:purchase], post)
end
```

I don't have strong opinions on this. If you do, or if you have a different spin that feels less redundant, I'm completely happy to go back to this style.

Reply:
Mostly thinking that often when adding options later, we need to condition on actions, and it would be less confusing then, but NBD.